roachtest: don't lose output in execCmd

tbg opened the pull request:
There was a race in which a command would finish and execCmdEx return
before the goroutines we had spawned to collect the output in a buffer
had finished doing so.

I was able to tickle the symptoms in #48476 before this commit by
adding a strategic `Sleep(time.Second)` in said goroutines, so
relatively confident that this:

Closes #48476.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/49371)
<!-- Reviewable:end -->

petermattis reviewed (commented):
Nice find!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49371#-:-M7rZ4zA28GHLAXZPrOe:bdkgtm2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @petermattis and @tbg)

___
*[pkg/cmd/roachtest/cluster.go, line 344 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49371#-M7rY4aJ4Pf1_l_WZeda:-M7rY4aJ4Pf1_l_WZedb:bn30opm) ([raw file](https://github.com/cockroachdb/cockroach/blob/302479dbcc61894535fd4ce377d01a012241aecc/pkg/cmd/roachtest/cluster.go#L344)):*
> ```Go
> // included in cmdRes.err.
> func execCmdEx(ctx context.Context, l *logger, args ...string) cmdRes {
> 	// NB: It is important that this waitgroup Waits after cancel() below.
> ```

Apparently it wasn't important.
___
*[pkg/cmd/roachtest/cluster.go, line 368 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49371#-M7rYPiv4vVOWfsWUwr8:-M7rYPiw553sZPa-G8lt:b-v0m0i8) ([raw file](https://github.com/cockroachdb/cockroach/blob/04df426ce759feffc4c376950dc4134cbceb857f/pkg/cmd/roachtest/cluster.go#L368)):*
> ```Go
> 		}
> 		defer rOut.Close()
> 		defer wOut.Close()
> ```

Is it problematic to close `wOut` (and `wErr`) twice? We'll close it in the defer but also when `closePipes()` is invoked. I suspect this isn't good. `wOut` has a file descriptor internally, and closing the file descriptor twice could cause the second close to accidentally close something completely unrelated to the pipe if the file descriptor was reused in between.

Rather than the `closePipes` closure, I think we want to have `wOut` and `wErr` variables. This defer then becomes:

```
defer func() {
  if wOut != nil {
    wOut.Close()
  }
}()
```

And when we want to close the pipes we do:

```
wOut.Close()
wOut = nil
wErr.Close()
wErr = nil
```


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49371)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @petermattis)

___
*[pkg/cmd/roachtest/cluster.go, line 344 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49371#-M7rY4aJ4Pf1_l_WZeda:-M7rZiyf7jpPBNtgxZuf:b-q2fo8e) ([raw file](https://github.com/cockroachdb/cockroach/blob/302479dbcc61894535fd4ce377d01a012241aecc/pkg/cmd/roachtest/cluster.go#L344)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Apparently it wasn't important.
</blockquote></details>

It stopped being important when the cancellation listener goroutine stopped being part of the wait group (which was always silly).
___
*[pkg/cmd/roachtest/cluster.go, line 368 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49371#-M7rYPiv4vVOWfsWUwr8:-M7rZsrbBNlR2Gp87SHe:b-7r5ibg) ([raw file](https://github.com/cockroachdb/cockroach/blob/04df426ce759feffc4c376950dc4134cbceb857f/pkg/cmd/roachtest/cluster.go#L368)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Is it problematic to close `wOut` (and `wErr`) twice? We'll close it in the defer but also when `closePipes()` is invoked. I suspect this isn't good. `wOut` has a file descriptor internally, and closing the file descriptor twice could cause the second close to accidentally close something completely unrelated to the pipe if the file descriptor was reused in between.

Rather than the `closePipes` closure, I think we want to have `wOut` and `wErr` variables. This defer then becomes:

```
defer func() {
  if wOut != nil {
    wOut.Close()
  }
}()
```

And when we want to close the pipes we do:

```
wOut.Close()
wOut = nil
wErr.Close()
wErr = nil
```
</blockquote></details>

Double closing returns an error (as per docs and the fact that these tests work), but is otherwise fine. I took your suggestion anyway (in spirit).


<!-- Sent from Reviewable.io -->

petermattis reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49371#-:-M7rccIRBVjNwKTJnHKO:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @petermattis)



<!-- Sent from Reviewable.io -->

tbg commented:
Ugh, just spent some time rediscovering that the deadline on the read pipes was necessary. I guess at least now we have test coverage for it.

bors r=petermattis

tbg commented:
bors r-

My bad

tbg committed:
roachtest: don't lose output in execCmd

There was a race in which a command would finish and execCmdEx return
before the goroutines we had spawned to collect the output in a buffer
had finished doing so.

I was able to tickle the symptoms in #48476 before this commit by
adding a strategic `Sleep(time.Second)` in said goroutines, so
relatively confident that this:

Closes #48476.

Release note: None

tbg commented:
bors r=petermattis

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49371)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @knz and @tbg)

___
*[pkg/cmd/roachtest/cluster.go, line 388 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49371#-M8M8wsR8a7VptUnBT48:-M8M8wsS4NGAQ6ZAdPNv:b62qqwd) ([raw file](https://github.com/cockroachdb/cockroach/blob/f9fea4d76a76d8a7a3a74331f21a7e59f8a71134/pkg/cmd/roachtest/cluster.go#L388)):*
> ```
> https://medium.com/@felixge/killing-a-child-process-and-all-of-its-children-in-go-54079af94773
> ```

That's what I implemented in Shakespeare, and it's portable enough (linux, bsd, macos, linux for windows).
However looking at your impl made me notice mine is incomplete :sweat:  - I'll do some homework to see if we can abstract this in a standalone lib.


<!-- Sent from Reviewable.io -->
